trivup/apps/OauthbearerOIDCApp.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+#
+
+# Copyright (c) 2021, Magnus Edenhill
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from trivup import trivup
+from http.server import BaseHTTPRequestHandler, HTTPServer
+
+from jwcrypto import jwk
+import python_jwt as jwt
+from datetime import datetime, timedelta
+
+import json
+import argparse
+import requests
+
+public_keys = []
+
+
+def generate_token(lifetime, valid=True):
+    private_pem = key.export_to_pem(private_key=True, password=None)
+    private_key = jwk.JWK.from_pem(private_pem)
+
+    payload = {
+        'exp': datetime.utcnow() + timedelta(days=0, seconds=lifetime),
+        'iat': datetime.utcnow(),
+        'iss': "issuer",
+        'sub': "subject",
+        'aud': 'api://default'
+    }
+    header = {
+        "kid": "abcdefg"
+    }
+
+    token = jwt.generate_jwt(payload, private_key, 'RS256',
+                             timedelta(seconds=lifetime), other_headers=header)
+    if not valid:
+        token += "invalid"
+
+    token_map = {"access_token": "%s" % token}
+    return token_map
+
+
+def generate_public_key():
+    global key
```

Comment:
I don't see the reason why this needs to be a global variable, it only seems to be used here in this method (and if it were a global it should have far more unique name).

It is vey seldom a good idea to use global variables.

trivup allows multiple Clusters to be instantiated separately, and each Cluster is isolated. And at the Cluster level, state should only be shared if absolutely needed, and in that case through querying the other Apps in that Cluster.
So, for this thing here we should keep all these keys on the WebServerHandler class instance object.

Reply:
Any reason these methods can't be methods on the WebServerHandler class? That's the only context they're used in.